### ISSUE TITLE
SmartLifecycle beans are not closed in @SpringBootTest

### ISSUE THREAD

**Issue by Venthe** [spring-projects/spring-boot#28312]
Hey, I'd really like to get some help.
TL;DR : `MessageListenerContainers` which were created in the previous contexts are still listening, grinding the tests to a halt.

```
// Major libraries
Spring Boot 2.5.4
maven.compiler.target 11
spring-cloud 2020.0.3
ibmmq 2.5.0
testcontainer-junit 1.15.3
testcontainer-kafka 1.15.3
```

**Setup**
`@KafkaListener` and `@JmsListener`; `@Testcontainers` on abstract `@SpringBootTest` with `@DynamicProperty` used to load data from `@Containers`

**What can be seen**
`MessageListenerContainers` created with the data loaded in `@DynamicProperty` are persisted between runs - before contexts loads (With/without `@DirtiesContext`) I can see that both KafkaListeners and JMSListeners are trying to connect to their respective servers with the ports from the "previous (docker) container" - previous `@DynamicContext`, and by consequence prevous context

Additionally, MessageListenerContainer from Kafka and JMS are not related, except they share only one interface worth mentioning, `SmartLifecycle`

**What I suspect**
`SmartLifecycle` is not destroyed/stopped on context refresh.

Please help? :)

**Comment by Venthe** [comment]
Clarification: After testing with the debugger, it seems that 'some' SmartLifecycle implementors have their .stop() called, but not KafkaListenerEndpointRegistry. Test is annotated with @DirtiesContext(classMode = AFTER_CLASS)

I still believe that the problem is on the core Spring side, as the same error is with JMS

**Comment by bclozel** [comment]
@kerkhofsd I think that what you're seeing is not a bug, since all beans are closed as expected. This issue what about beans not being closed at all.

In this case, what you're seeing is the [Context Caching feature in action](https://docs.spring.io/spring-framework/docs/current/reference/html/testing.html#testcontext-ctx-management-caching): contexts are cached and reused within a test suite and are closed at the end.

**Comment by kerkhofsd** [comment]
@bclozel thanks for your feedback! 

I can see how this behaviour fits in the context caching feature. For `MessageListenerContainers` beans this does give some undesired side effects though. However indeed probably unrelated to this issue. 
Thanks again for your input!

**Comment by NicoStrecker** [comment]
@bclozel I have the same Problem after upgrading to Spring Boot 3.1.2 and switiching from ActiveMQ Classic to ActiveMQ Artemis.

`Failed to shut down 1 bean with phase value 2147483647 within timeout of 10000ms: [org.springframework.jms.config.internalJmsListenerEndpointRegistry]`

My Test uses `@DirtiesContext` (no Context Caching)

**Comment by outsideMyBox** [comment]
For what it is worth, I have the same issue as @bclozel in my tests (without any `@DirtyContext`)
`Shutdown phase 2147483647 ends with 1 bean still running after timeout of 30000ms: [org.springframework.jms.config.internalJmsListenerEndpointRegistry]`

Spring Boot 3.2.6
org.apache.activemq:artemis-junit 2.12.0

**Comment by bclozel** [comment]
@outsideMyBox I'm not seeing any issue on my side. If you believe you have found a bug, please create a new issue with a minimal application showing the problem.

**Comment by outsideMyBox** [comment]
Sorry, wrong person.

**Comment by fabian-flechtmann** [comment]
We ran into the same problem recently. Our SQS listeners were still actively listening from the Spring bean cache.

Our tests were flaky because the result depended on which listener picked up each message, which was random. If the cached listener picked it up, the stack traces made it look as if the mocking had failed or the message was lost.

Would it make sense, given that there seem to be similar issues with other listeners, to extend the SmartLifecycle interface?

Something like `onEnterContext` and `onLeaveContext` could be intercepted in the listener beans to toggle the listening. These methods would be called when a bean moves into or out of the active application context. It's not the same as `start` and `stop`, which are only called when the bean is loaded for the first time and when the application shuts down.

**Comment by wilkinsona** [comment]
@fabian-flechtmann that's an interesting suggestion but this is the wrong issue tracker for it. Any such change would have to be made in Spring Framework as there's where `SmartLifecycle` is maintained.

**Comment by fabian-flechtmann** [comment]
This has now been fixed in the spring-framework project ([#35168](https://github.com/spring-projects/spring-framework/issues/35168)) and will be included in 7.0 M7.

The fix introduces calling `stop()` when an application context is cached and becomes inactive, then calling `start()` again when it is reused. Listeners need to support multiple successive calls to these methods but i would expect this to already be the case.
